btrfs: add all subvolume commands

aplanas opened the pull request:
### What does this PR do?
Add all subvolume commands to the btrfs module

### Tests written?

Yes

aplanas committed:
btrfs: add all subvolume commands

isbm commented on salt/modules/btrfs.py:
```diff
@@ -673,3 +673,465 @@ def properties(obj, type=None, set=None):
+    res = __salt__['cmd.run_all'](cmd)
+    salt.utils.fsutils._verify_run(res)
+
+    lines = res['stdout'].split('\n')
```
```python
import os
...
....split(os.linesep)
```

aplanas replied:
I moved to ''.splitlines()

isbm commented on tests/unit/modules/test_btrfs.py:
```diff
@@ -362,3 +362,300 @@ def test_properties_error(self):
+        Test btrfs subvolume create
+        '''
+        # Fails when qgroupids is not a list
+        self.assertRaises(CommandExecutionError, btrfs.subvolume_create,
```
Please use pytest instead. And for the raising exception should be another UT.

aplanas replied:
Done

cachedout replied:
@isbm What is the motivation for just suddenly switching to pytest here? We certainly have long-term plans to go in that direction but are we sure we want to just start importing pytest into individual tests like this?

aplanas replied:
IMHO I do not like pytest. Basically because the magic that involves `assert`. In normal Python `assert` is a language statement that evaluate the expression (True or False) and throws an `AssertionError`.

Pytest change that to include the features that are already in the `self.assertEqual` family. Both code generate very similar report to show the differences of data structures. So I do not see the gain incorporating more magic to this problem.

But I am happy to adhere to the project expectations in regards to tests frameworks.

isbm replied:
> IMHO I do not like pytest.

`¯\_(ツ)_/¯`. This is what we are using here and pytest gives _better results_ to catch the error.

P.S. I like pytest.

isbm commented on tests/unit/modules/test_btrfs.py:
```diff
@@ -362,3 +362,300 @@ def test_properties_error(self):
+            'cmd.run_all': MagicMock(return_value={'recode': 0}),
+        }
+        with patch.dict(btrfs.__salt__, salt_mock):
+            self.assertTrue(btrfs.subvolume_create('var', dest='/mnt'))
```
Please use just plain `assert` from Python.

aplanas replied:
done

isbm commented on tests/unit/modules/test_btrfs.py:
```diff
@@ -362,3 +362,300 @@ def test_properties_error(self):
+
+        # Parameter commit can be 'after' or 'each'
+        self.assertRaises(CommandExecutionError, btrfs.subvolume_delete,
+                          name='var', commit='maybe')
```
Please break them into three UT

aplanas replied:
I moved into two, as the two first are testing basically the same.

isbm reviewed (changes requested):
Please take a look at these points and please use pytest's ecosystem instead.

isbm commented on salt/modules/btrfs.py:
```diff
@@ -673,3 +673,465 @@ def properties(obj, type=None, set=None):
+    cmd = ['btrfs', 'subvolume', 'delete']
+    if commit == 'after':
+        cmd.append('--commit-after')
+    if commit == 'each':
```
`elif`

aplanas replied:
done

isbm commented on salt/modules/btrfs.py:
```diff
@@ -673,3 +673,465 @@ def properties(obj, type=None, set=None):
+            line = ''
+        else:
+            value, line = line.split(' ', 1)
+    return (value, line.strip())
```
Extra parenthesis

isbm replied:
Or so:
```python
value = None
if line.startswith(key):
    line = line[len(key):].strip()
    value, line = use_rest and (line, '') or line.split(' ', 1)
return value, line.strip()

aplanas replied:
Why to obfuscate the code with `and`s and `or`s like if this is pre-Python 2.4?

I removed the extra line with `.strip()` and removed the optional parents (I usually like to be explicit when I return a tuple, but I guess that here is clear enough)

isbm commented on salt/modules/btrfs.py:
```diff
@@ -673,3 +673,465 @@ def properties(obj, type=None, set=None):
+    if dest and name:
+        name = os.path.join(dest, name)
+    if name:
+        cmd.append(name)
```
No need in all these `if if if`:
```python
cmd.append(os.path.join(dest or '', name or '').strip(os.path.sep))
```

aplanas replied:
I am not sure why to favor a more obfuscated code.

I am personally not so smart to realize that the last `.strip(os.path.sep)` is needed because if `name` is not assigned, `os.path.join(dest, '')` will add this unneeded extra '/'.

Let me be easy and explicit here!

aplanas committed:
btrfs: add subvolume_exists

aplanas committed:
btrfs: create and delete return False if noop

If the create and delete actions are not needed, the function will
return False. This indicate that the submodule was already there,
or was missing on the first place.

aplanas commented:
I changed the code a bit. I added one new function, `subvolume_exists`, and I changed `create` and `delete` to first check if the volume is there or is missing.

I updated the tests and added more scenarios to cover the new features.

aplanas commented on salt/modules/btrfs.py:
```diff
@@ -673,3 +673,496 @@ def properties(obj, type=None, set=None):
+
+    res = __salt__['cmd.run_all'](cmd)
+    salt.utils.fsutils._verify_run(res)
+    return True
```
`salt.utils.fsutils._verify_run` throws an exception then the command fails and log the reason.

Do you think that makes more sense to catch the CommandExecutionError and return False there?

isbm replied:
No, I just saw you return just `True`, so I am wondering if that doesn't depend on how we are going to use it. E.g. if this boolean is going to be reused something like:

```jinja2
{%if __salt__['btrfs.subvolume_create'](...)%}
   {# Do something #}
{% else %}
   {# Do something else #}
{% endif %}
```

aplanas replied:
Yeah, puzzle me why we throw exceptions.

In any case subvolume_create and delete are aware of the existence (or not) of the subvolume:

https://github.com/saltstack/salt/pull/50541/files/14004026ad048a20badd4c68f79c43360834c757#diff-af9d04adc25a92fdf8fece1a0762dad7R733

So will return True if the subvolume was already there or was created later. A fail is that something went very bad with the command itself.

isbm replied:
But I thought maybe it is correct to return `False` if subvol already existed? Because your function is called "subvolume_**create**".

aplanas replied:
I really do not know!

I do not mind to change it if is more expressive, but I wonder if a False can wrongly suggest that the submodule was not created, even if was already there.

isbm replied:
Well, I think this is fine, apparently. I hear your point for wrong suggest, but apparently why not: it says _create_. So was it _created_? — no, as it was no-op. Therefore clearly `False`. To know if it exists or not, should be _exists_ function. And so:

```jinja2
{% if not subvol_create('blah') and not subvol_exists('blah') %}
  <h>SCREAM!</h>
{% endif %}
```

isbm replied:
OK, I think this is good. Except maybe update the documentation so we know what it returns when. Other then that we're good to go.

aplanas replied:
Documented.

aplanas commented on salt/modules/btrfs.py:
```diff
@@ -673,3 +673,496 @@ def properties(obj, type=None, set=None):
+
+    res = __salt__['cmd.run_all'](cmd)
+    salt.utils.fsutils._verify_run(res)
+    return True
```
same answer

aplanas commented:
I changed the `subvolume_exists` function to avoid logging ERROR in case that a subvolume do not exits, but functionally is equivalent. Also I added the documentation required.

cachedout commented:
Merged! Thanks, @aplanas
